simulatingrisk/hawkdovemulti/model.py

```diff
@@ -194,6 +212,52 @@ def __init__(
         # store whether to compare cumulative payoff or since last adjustment round
         self.adjust_payoff = adjust_payoff
 
+    def _risk_level_in_bounds(self, value):
+        # check if a generated risk level is within bounds
+        return self.min_risk_level <= value <= self.max_risk_level
+
+    def get_risk_attitude_generator(self):
+        """return a generator that will return risk attitudes for individual
+        agents based on the configured distribution."""
+        if self.risk_distribution == "uniform":
+            # uniform/random: generate random integer within risk level range
+            while True:
+                yield self.random.randint(self.min_risk_level, self.max_risk_level)
+        if self.risk_distribution == "normal":
+            # return values from a normal distribution centered around 4
+            while True:
+                yield int(self.random.gauss(4, 1.5))
+        elif self.risk_distribution == "skewed left":
+            # return values from a triangler distribution centered around 0
+            while True:
+                yield round(
+                    self.random.triangular(self.min_risk_level, self.max_risk_level, 0)
+                )
+        elif self.risk_distribution == "skewed right":
+            # return values from a triangler distribution centered around 8
+            while True:
+                yield round(
+                    self.random.triangular(self.min_risk_level, self.max_risk_level, 8)
+                )
+        elif self.risk_distribution == "bimodal":
+            # to generate a bimodal distribution, alternately generate
+            # values from two different normal distributions centered
+            # around the beginning and end of our risk attitude range
+            while True:
+                yield int(self.random.gauss(1, 0.75))
+                yield int(self.random.gauss(7, 0.75))
+
+    def get_risk_attitude(self):
+        """return the next value from risk attitude generator, based on
+        configured distribution."""
+        val = next(self.risk_attitude_generator)
+        # occasionally generators will return values that are out of range.
+        # rather than capping to the min/max and messing up the distribution,
+        # just get the next value
+        while not self._risk_level_in_bounds(val):
+            val = next(self.risk_attitude_generator)
+        return val
```

Comment:
so this is what keeps those `while True`'s above from infinite-looping? nice

Reply:
Yeah, was looking for a good solution for alternating values for the bimodal and figured out I could return a generator for all of them that generates however many values you need.   I was initially looking at the distribution options in numpy, but couldn't figure out a way to do anything like this - you have to generate all the values at once based on the needed size.

Reply:
With the bimodal distribution centered around 7, I almost never got a value of 9 on the initial distribution ("step 1"), though I did get a value of 5 quite frequently. The probability of these should be symmetric, right (aside from the miniscule probability of the distribution centered around 1 yielding 5)? Might want to check to make sure that the range ends at 9, rather than 8, everywhere? (Though I did occassionally get a value of 9, so I'm not totally sure what's happening.)